### PR TITLE
Fix compilation for boost 1.74

### LINKED ISSUE #3171: Compiling fails with boost 1.74
Bug Report
==========
It seems that there are some changes in the boost library. 
Cmake is warning about this. So it's more a todo then a bug. 

freeorion/GG/src/Texture.cpp:244:86:   required from here
/usr/include/boost/gil/extension/dynamic_image/any_image.hpp:53:52: error: 'struct boost::mp11::mp_list<boost::gil::image<boost::gil::pixel<unsigned char, boost::gil::layout<boost::mp11::mp_list<boost::gil::gray_color_t>, boost::mp11::mp_list<std::integral_constant<int, 0> > > >, false, std::allocator<unsigned char> >, boost::gil::image<boost::gil::pixel<unsigned char, boost::gil::layout<boost::mp11::mp_list<boost::gil::gray_color_t, boost::gil::alpha_t>, boost::mp11::mp_list<std::integral_constant<int, 0>, std::integral_constant<int, 1> > > >, false, std::allocator<unsigned char> >, boost::gil::image<boost::gil::pixel<unsigned char, boost::gil::layout<boost::mp11::mp_list<boost::gil::red_t, boost::gil::green_t, boost::gil::blue_t>, boost::mp11::mp_list<std::integral_constant<int, 0>, std::integral_constant<int, 1>, std::integral_constant<int, 2> > > >, false, std::allocator<unsigned char> >, boost::gil::image<boost::gil::pixel<unsigned char, boost::gil::layout<boost::mp11::mp_list<boost::gil::red_t, boost::gil::green_t, boost::gil::blue_t, boost::gil::alpha_t>, boost::mp11::mp_list<std::integral_constant<int, 0>, std::integral_constant<int, 1>, std::integral_constant<int, 2>, std::integral_constant<int, 3> > > >, false, std::allocator<unsigned char> > >' has no member named 'recreate'
   53 |     result_type operator()(Image& img) const { img.recreate(_dimensions,_alignment); }

Environment
-----------
Linux: Gentoo ~amd64
gcc (Gentoo 10.2.0-r1 p2) 10.2.0
Python 3.7.9
boost 1.74

* **FreeOrion Version**:
Building v0.4.10+ build 2020-09-16.4ce1758

* **Operating System**:
Gentoo Linux
* Compiled from source

Description
-----------
compiling failed 

Expected Result
---------------
compile freeorion

Steps to reproduce
------------------
mkdir build 
cd build 
cmake ..
make 

Kind regards Bernd Wollny

### PR DESCRIPTION
**Pull request by o01eg**
Fixes #3171

### CODE REVIEW

File: GG/CMakeLists.txt [hunk 1]
```diff
@@ -122,6 +122,7 @@ if(NOT BUILD_HEADLESS)
             $<$<VERSION_LESS:${Boost_MAJOR_VERSION}.${Boost_MINOR_VERSION},1.68>:GIGI_CONFIG_USE_OLD_IMPLEMENTATION_OF_GIL_PNG_IO>
             $<$<VERSION_GREATER:${Boost_MAJOR_VERSION}.${Boost_MINOR_VERSION},1.67>:BOOST_GIL_IO_ENABLE_GRAY_ALPHA>
             $<$<VERSION_GREATER:${Boost_MAJOR_VERSION}.${Boost_MINOR_VERSION},1.71>:BOOST_GIL_USES_MP11>
+            $<$<VERSION_GREATER:${Boost_MAJOR_VERSION}.${Boost_MINOR_VERSION},1.73>:BOOST_GIL_USES_VARIANT2>
```
Comment (geoffthemedio) [inline GG/CMakeLists.txt]: doesn't work for MSVC...

File: GG/src/Texture.cpp [hunk 2]
```diff
@@ -24,7 +24,9 @@
 #  include <boost/gil/extension/io/png.hpp>
 # endif
 #endif
-#if BOOST_VERSION >= 107000
+#if defined (BOOST_GIL_USES_VARIANT2)
```
Comment (geoffthemedio) [inline GG/src/Texture.cpp]: suggest adding before this
````
#ifndef BOOST_GIL_USES_VARIANT2
#  if BOOST_VERSION >= 107400
#    define BOOST_GIL_USES_VARIANT2
#  endif
#endif
````

File: GG/CMakeLists.txt (hunk 1 above)
Comment (o01eg) [reply GG/CMakeLists.txt]: MSVC should use boost from SDK

Comment (geoffthemedio) [reply GG/CMakeLists.txt]: SDK currently has 1.70 (or at least my version does), but it should be compatible with future Boost version updates.

File: GG/src/Texture.cpp (hunk 2 above)
Comment (geoffthemedio) [reply GG/src/Texture.cpp]: can also wrap in `#ifdef _MSC_VER`

File: GG/CMakeLists.txt (hunk 1 above)
Comment (o01eg) [reply GG/CMakeLists.txt]: I suppose Freeorion MSVC project updates when switch to new SDK.

Comment (geoffthemedio) [reply GG/CMakeLists.txt]: probably, but I'd like to future-proof it now while we're looking at it.

Comment (o01eg) [reply GG/CMakeLists.txt]: Maybe then we shouldn't use custom definitions based on boost version and use `BOOST_VERSION` in sources instead? Or define them somewhere in `config.h` file.

Comment (o01eg) [reply GG/CMakeLists.txt]: I've removed definition from `CMakeLists.txt`.

File: GG/src/Texture.cpp (hunk 2 above)
Comment (o01eg) [reply GG/src/Texture.cpp]: I've used `BOOST_VERSION` here too.

**Comment by geoffthemedio** [comment]
seems ok

**Pull request merged**